### PR TITLE
Fix html repr on non-str keys

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] Closes https://github.com/pydata/xarray/issues/3833
 - [x] Tests added
 - [x] Passes `isort -rc . && black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by max-sixty** [comment]
Docs [failing](https://dev.azure.com/xarray/xarray/_build/results?buildId=2380&view=logs&j=7e620c85-24a8-5ffa-8b1f-642bc9b1fc36&t=68484831-0a19-5145-bfe9-6309e5f7691d&l=381), I think on:

`/usr/share/miniconda/envs/xarray-tests/lib/python3.8/site-packages/sphinx_rtd_theme/search.html:21: RemovedInSphinx30Warning: To modify script_files in the theme is deprecated. Please insert a <script> tag directly in your theme instead.`

Anyone have any ideas?

**Comment by keewis** [comment]
This warning is something we can't do anything about; we have to wait for a release of `sphinx_rtd_theme` (it is fixed on their master).

It does not cause sphinx to fail, though. In general, if the docs are failing it's mostly due to some error / warning elevated to error in the "reading" step, in this [case](https://dev.azure.com/xarray/xarray/_build/results?buildId=2380&view=logs&j=7e620c85-24a8-5ffa-8b1f-642bc9b1fc36&t=68484831-0a19-5145-bfe9-6309e5f7691d&l=276):
```
/home/vsts/work/1/s/doc/whats-new.rst:91: WARNING: Inline literal start-string without end-string.
```
which should be easy to fix (I assume using "`Dataset`s" without a space after the backtick is forbidden?).

File: doc/whats-new.rst [hunk 1]
```diff
@@ -88,6 +88,8 @@ Bug fixes
   
 - Fix :py:meth:`xarray.core.dataset.Dataset.to_zarr` when using `append_dim` and `group`
   simultaneously. (:issue:`3170`). By `Matthias Meyer <https://github.com/niowniow>`_.
+- Fix html repr on ``Dataset``s with non-string keys (:pull:`3807`).
```
Comment (keewis) [inline doc/whats-new.rst]:
```suggestion
- Fix html repr on :py:class:`Dataset` with non-string keys (:pull:`3807`).
```

**Comment by max-sixty** [comment]
Thanks @keewis !

**Pull request merged**